### ISSUE TITLE
Make it impossible for players to get "stuck"

### ISSUE THREAD

**Issue by telenc** [GallagherAiden/footballSimulationEngine#33]
Hello,
I wanted to test with a team perso.
My players go directly to their camps, next to the goalkeeper and nothing happens ... the ball always goes to the bottom right side and nothing happens.
What is my problem?

Thank you

Team 1
```
{
  "name": "no",
  "manager": "manager no",
  "players": [
    {
      "name": "Rodrick Helmy",
      "position": "GK",
      "rating": 77,
      "injured": false,
      "startPOS": [
        60,
        0
      ],
      "skill": {
        "passing": 81,
        "shooting": 68,
        "tackling": 68,
        "saving": 18,
        "agility": 75,
        "strength": 84,
        "penalty_taking": 68,
        "jumping": 180
      },
      "fitness": 100
    },
    {
      "name": "Antwan Rousselle",
      "position": "LB",
      "rating": 77,
      "injured": false,
      "startPOS": [
        30,
        20
      ],
      "skill": {
        "passing": 80,
        "shooting": 68,
        "tackling": 68,
        "saving": 16,
        "agility": 74.5,
        "strength": 83,
        "penalty_taking": 68,
        "jumping": 160
      },
      "fitness": 100
    },
    {
      "name": "Malcom Haessig",
      "position": "CB",
      "rating": 76,
      "injured": false,
      "startPOS": [
        50,
        60
      ],
      "skill": {
        "passing": 81,
        "shooting": 65,
        "tackling": 68,
        "saving": 19,
        "agility": 75,
        "strength": 82,
        "penalty_taking": 65,
        "jumping": 190
      },
      "fitness": 100
    },
    {
      "name": "Rico Vornes",
      "position": "CB",
      "rating": 77,
      "injured": false,
      "startPOS": [
        50,
        20
      ],
      "skill": {
        "passing": 82,
        "shooting": 65,
        "tackling": 51,
        "saving": 19,
        "agility": 81.5,
        "strength": 68,
        "penalty_taking": 65,
        "jumping": 190
      },
      "fitness": 100
    },
    {
      "name": "Mikel Emberling",
      "position": "RB",
      "rating": 78,
      "injured": false,
      "startPOS": [
        90,
        20
      ],
      "skill": {
        "passing": 83,
        "shooting": 68,
        "tackling": 53,
        "saving": 15,
        "agility": 81.5,
        "strength": 65,
        "penalty_taking": 68,
        "jumping": 150
      },
      "fitness": 100
    },
    {
      "name": "Russell Walmer",
      "position": "LM",
      "rating": 75,
      "injured": false,
      "startPOS": [
        30,
        120
      ],
      "skill": {
        "passing": 69,
        "shooting": 80,
        "tackling": 50,
        "saving": 17,
        "agility": 68,
        "strength": 80,
        "penalty_taking": 80,
        "jumping": 170
      },
      "fitness": 100
    },
    {
      "name": "Bernardo Balent",
      "position": "CAM",
      "rating": 74,
      "injured": false,
      "startPOS": [
        50,
        140
      ],
      "skill": {
        "passing": 68,
        "shooting": 52,
        "tackling": 54,
        "saving": 81,
        "agility": 50,
        "strength": 66,
        "penalty_taking": 52,
        "jumping": 181
      },
      "fitness": 100
    },
    {
      "name": "Isreal Lebaugh",
      "position": "CDM",
      "rating": 79,
      "injured": false,
      "startPOS": [
        50,
        100
      ],
      "skill": {
        "passing": 80,
        "shooting": 54,
        "tackling": 82,
        "saving": 15,
        "agility": 59,
        "strength": 82,
        "penalty_taking": 54,
        "jumping": 150
      },
      "fitness": 100
    },
    {
      "name": "Demetrius Colom",
      "position": "RM",
      "rating": 79,
      "injured": false,
      "startPOS": [
        90,
        120
      ],
      "skill": {
        "passing": 81,
        "shooting": 50,
        "tackling": 84,
        "saving": 17,
        "agility": 58.5,
        "strength": 80,
        "penalty_taking": 50,
        "jumping": 170
      },
      "fitness": 100
    },
    {
      "name": "Caleb Fukui",
      "position": "ST",
      "rating": 79,
      "injured": false,
      "startPOS": [
        50,
        270
      ],
      "skill": {
        "passing": 83,
        "shooting": 51,
        "tackling": 83,
        "saving": 16,
        "agility": 59,
        "strength": 81,
        "penalty_taking": 51,
        "jumping": 160
      },
      "fitness": 100
    },
    {
      "name": "Romeo Pelch",
      "position": "ST",
      "rating": 76,
      "injured": false,
      "startPOS": [
        50,
        280
      ],
      "skill": {
        "passing": 81,
        "shooting": 67,
        "tackling": 68,
        "saving": 17,
        "agility": 74,
        "strength": 82,
        "penalty_taking": 67,
        "jumping": 170
      },
      "fitness": 100
    }
  ]
}
``` 

```
{
  "name": "no",
  "manager": "manager no2",
  "players": [
    {
      "name": "Rory Roya",
      "position": "GK",
      "rating": 73,
      "injured": false,
      "startPOS": [
        60,
        0
      ],
      "skill": {
        "passing": 66,
        "shooting": 50,
        "tackling": 53,
        "saving": 82,
        "agility": 51.5,
        "strength": 65,
        "penalty_taking": 50,
        "jumping": 182
      },
      "fitness": 100
    },
    {
      "name": "Isaias Apa",
      "position": "LB",
      "rating": 79,
      "injured": false,
      "startPOS": [
        30,
        20
      ],
      "skill": {
        "passing": 83,
        "shooting": 52,
        "tackling": 83,
        "saving": 15,
        "agility": 60,
        "strength": 80,
        "penalty_taking": 52,
        "jumping": 150
      },
      "fitness": 100
    },
    {
      "name": "Irving Stephanie",
      "position": "CB",
      "rating": 79,
      "injured": false,
      "startPOS": [
        50,
        40
      ],
      "skill": {
        "passing": 83,
        "shooting": 51,
        "tackling": 82,
        "saving": 16,
        "agility": 58.5,
        "strength": 81,
        "penalty_taking": 51,
        "jumping": 160
      },
      "fitness": 100
    },
    {
      "name": "Cody Ahl",
      "position": "CB",
      "rating": 81,
      "injured": false,
      "startPOS": [
        50,
        20
      ],
      "skill": {
        "passing": 84,
        "shooting": 52,
        "tackling": 84,
        "saving": 17,
        "agility": 59,
        "strength": 83,
        "penalty_taking": 52,
        "jumping": 170
      },
      "fitness": 100
    },
    {
      "name": "Ivory Bornmann",
      "position": "RB",
      "rating": 78,
      "injured": false,
      "startPOS": [
        90,
        20
      ],
      "skill": {
        "passing": 81,
        "shooting": 54,
        "tackling": 80,
        "saving": 16,
        "agility": 58.5,
        "strength": 82,
        "penalty_taking": 54,
        "jumping": 160
      },
      "fitness": 100
    },
    {
      "name": "Freddie Basco",
      "position": "LM",
      "rating": 77,
      "injured": false,
      "startPOS": [
        30,
        120
      ],
      "skill": {
        "passing": 84,
        "shooting": 68,
        "tackling": 68,
        "saving": 16,
        "agility": 74.5,
        "strength": 80,
        "penalty_taking": 68,
        "jumping": 160
      },
      "fitness": 100
    },
    {
      "name": "Milton Carolina",
      "position": "CAM",
      "rating": 77,
      "injured": false,
      "startPOS": [
        50,
        140
      ],
      "skill": {
        "passing": 84,
        "shooting": 67,
        "tackling": 66,
        "saving": 15,
        "agility": 73,
        "strength": 83,
        "penalty_taking": 67,
        "jumping": 150
      },
      "fitness": 100
    },
    {
      "name": "Deshawn Rugama",
      "position": "CDM",
      "rating": 77,
      "injured": false,
      "startPOS": [
        50,
        100
      ],
      "skill": {
        "passing": 83,
        "shooting": 67,
        "tackling": 66,
        "saving": 16,
        "agility": 73,
        "strength": 84,
        "penalty_taking": 67,
        "jumping": 160
      },
      "fitness": 100
    },
    {
      "name": "Duncan Beske",
      "position": "RM",
      "rating": 78,
      "injured": false,
      "startPOS": [
        90,
        120
      ],
      "skill": {
        "passing": 83,
        "shooting": 68,
        "tackling": 66,
        "saving": 17,
        "agility": 75,
        "strength": 84,
        "penalty_taking": 68,
        "jumping": 170
      },
      "fitness": 100
    },
    {
      "name": "Darren Moronta",
      "position": "ST",
      "rating": 79,
      "injured": false,
      "startPOS": [
        50,
        280
      ],
      "skill": {
        "passing": 84,
        "shooting": 67,
        "tackling": 50,
        "saving": 15,
        "agility": 83.5,
        "strength": 65,
        "penalty_taking": 67,
        "jumping": 150
      },
      "fitness": 100
    },
    {
      "name": "Thurman Bowcock",
      "position": "ST",
      "rating": 75,
      "injured": false,
      "startPOS": [
        50,
        270
      ],
      "skill": {
        "passing": 65,
        "shooting": 83,
        "tackling": 54,
        "saving": 15,
        "agility": 68,
        "strength": 80,
        "penalty_taking": 83,
        "jumping": 150
      },
      "fitness": 100
    }
  ]
}
```

<img width="1056" alt="capture d ecran 2019-02-04 a 11 25 22" src="https://user-images.githubusercontent.com/5463621/52202763-be422e80-286f-11e9-94db-6c1534cc85c6.png">

---- UPDATE ----
I had jumping below 100. I modified but it does not change anything

**Comment by GallagherAiden** [comment]
Sorry you’re seeing this, Can I ask what version of the footballsimulationengine you are using? I thought I had fixed this is 2.1.1, I will try to recreate and publish a fix asap

**Comment by telenc** [comment]
Hello !
I cloned and copy pasted this github.

```
  "version": "2.1.0",
```

Thank's !

**Comment by GallagherAiden** [comment]
Hi @telenc, I've managed to recreate the issue and will work on making it not possible to get stuck in these positions in this way.

Based on 20 different games, it is less likely for the players to get stuck in these positions if the pitch is made larger. 
Could you try again with the pitch and team data in: 
https://github.com/GallagherAiden/footballSimulationEngine/tree/master/init_config

 - pitch.json
 - team1.json
 - team2.json

That **should** make a difference. And I will work on making it impossible for players to get stuck in positions

**Comment by telenc** [comment]
Thank you for taking the time to help me.
I would like to use your tools for my football simulation game. the data of the teams that I gave are data coming from my game. The problem is that I need to test with this data. I tested with different pitches. I admit that at moment I have a "good" match  but at sometimes it bug a lot: /. So I can not integrate it into my game because if a match on two remains blocked I may have a lot of 0-0.
I would like to help you I tried to look at my side but I admit that it's complicated to debug this kind of tool when we do not know every code! There is already a big job on it.

Would you have a techinque to debug this kind of behavior? I would like to help you

I join you a video that I could do. The first match the players remains stuck at the commitment. The second match starts to be blocked at the level of 5000 iterations

Thank you for taking time
Congratulations for your tools I hope you can do a complete trick because it is a great tools


https://youtu.be/QSc5_5vgivE

**Comment by telenc** [comment]
@GallagherAiden  Hello ! i have found a problem ... 
But I did not find how to fix it for the moment ...
I added logs. which shows me the player who has the "hasBall" parameter = true
so he shows me the player "Romeo Pelch"

The problem is that in the function makeMovement you have a condition
"if (player.hasBall) {"
he must necessarily go towards the opponent's goal!
the trouble is that the ball is behind him! so he goes forward, and after he goes back (I did not find where he is going back)

I think there is a problem with the variable "hasBall" it is not well set on the player ...: /

![romeooo](https://user-images.githubusercontent.com/5463621/52373729-adeea700-2a5b-11e9-9d64-bdc15acda04d.png)
<img width="1030" alt="capture d ecran 2019-02-06 a 22 03 56" src="https://user-images.githubusercontent.com/5463621/52373734-b0e99780-2a5b-11e9-9186-2f8ffded01f9.png">

**Comment by GallagherAiden** [comment]
hi @telenc, So I've put a temporary fix into this github to unblock the issue of players getting stuck. (I haven't added it to the node module because I want to find what is actually causing the issue.) you can find the commit here - it should stop the players getting 'stuck' for too long. I'll keep looking into fixing the issue.
https://github.com/GallagherAiden/footballSimulationEngine/commit/7bf5cf846b752f52049abe1fffa6f23047179121

**Comment by GallagherAiden** [comment]
basically, is the players get stuck I set a freekick for 'time wasting'. I don't like this as a long term solution, but it will make your game more playable.

We see less of the 'stuck' state if the pitch is larger - these are the statistics for amount of time wasting free kicks given:
Time Wasting Output

Run 1 - 70 time wasting free kicks (6000 its, 6000 its) [2 - 2] pitchSize = { "pitchWidth": 120, "pitchHeight": 600  }
Run 2 - 38 time wasting free kicks (6000 its, 6000 its) [2 - 3] pitchSize = { "pitchWidth": 120, "pitchHeight": 600  }
Run 3 - 52 time wasting free kicks (6000 its, 6000 its) [3 - 4] pitchSize = { "pitchWidth": 120, "pitchHeight": 600  }
Run 4 - 67 time wasting free kicks (6000 its, 6000 its) [1 - 3] pitchSize = { "pitchWidth": 120, "pitchHeight": 600  }
Run 5 - 10 time wasting free kicks (6000 its, 6000 its) [1 - 1] pitchSize = { "pitchWidth": 680, "pitchHeight": 1050 }
Run 6 - 11 time wasting free kicks (6000 its, 6000 its) [1 - 2] pitchSize = { "pitchWidth": 680, "pitchHeight": 1050 }
Run 7 - 12 time wasting free kicks (6000 its, 6000 its) [2 - 3] pitchSize = { "pitchWidth": 680, "pitchHeight": 1050 }
Run 8 - 9 time wasting free kicks (6000 its, 6000 its) [2 - 1] pitchSize = { "pitchWidth": 680, "pitchHeight": 1050 }

**Comment by GallagherAiden** [comment]
"if (player.hasBall) {" - this bug is interesting. I'll take a look at what could be causing this :)

**Comment by telenc** [comment]
Thank's it's better :)

**Comment by GallagherAiden** [comment]
@telenc - I looked into the issue of players set to 'hasBall' when the ball is not with them.

https://github.com/GallagherAiden/footballSimulationEngine/commit/47520d9c3f05288040250f89d675fc0d3a085961 

Please find the commit. This will be included in the next npm version (v2.1.2+) for now you'll need to pull from the github directly.

Players can be deemed to 'have the ball' between -5 and 5 in the X and Y position - this is because its simulates tackling scenarios, kicking the ball slightly forward etc.

Is there anything else I can help with?

**Comment by telenc** [comment]
Hi ! Thank you very much for taking into account super fast these remarks! I look in my corner :) if I see bug I do the same I look a little bit what it is if I manage to settle why not otherwise I inform you of my remarks! Thank you very much I test all that!

**Comment by telenc** [comment]
After a first test ... bravo! it starts to look like a real match it's great!
There is still some concern about shooting (see message above). The passes that sometimes go directly into touch when a player is right next (basically they do not go in the right direction) or do all the ground (for nothing because there is nobody on the other side, can it's a shot I can not see the difference of the shot)

There is still a little worry about hasball, indeed sometimes I have a player who has the ball while a team is right next to the ball!

I would look on my side if I can not or is the worry
I would like to add a system of player change when the player is too tired, I will try to make you a pull request when I finish (because in my game there is a system of replacement)

Thank you in any case to take into account my worries quickly!

**Comment by GallagherAiden** [comment]
@telenc 
So player1 is closer to the ball but another player (player2) has the ball? 
   I can take another look into this :) It relates to this open issue: https://github.com/GallagherAiden/footballSimulationEngine/issues/28
    
Can't see a comment about 'shooting', what issue are we seeing? Happy to take a look

**Comment by GallagherAiden** [comment]
@telenc - I looked into the Closest Player to Ball and have found that in the 'playIteration' function I am doing this:
- find team1 closest player
- find team2 closest player
- moveball
- decideMovement
.... obviously this causes errors! because the ball moves AFTER i find the closest player. https://github.com/GallagherAiden/footballSimulationEngine/commit/d10b81d863e7b1ec941f255c0b4b04b91976f792

I have fixed this in this commit and added some test cases. New process:
- moveball
- find team1 closest player
- find team2 closest player
- decideMovement

**Comment by GallagherAiden** [comment]
@telenc - I think I found what was making the players 'stuck'. It was a mixture of bad passing, the ball not 'moving' with the players. I've made the relevant changes https://github.com/GallagherAiden/footballSimulationEngine/commit/3b750a276ef464c6a7a92ae8a4075190721b7d4a and have removed the 'time wasting' 'staticCount' check.

Could you retest your issues on the current Github version of the code?

**Comment by GallagherAiden** [comment]
I've also made changes to https://github.com/GallagherAiden/footballsimulationexample which now has up to date examples

**Comment by GallagherAiden** [comment]
Not observed in latest commit: https://github.com/GallagherAiden/footballSimulationEngine/commit/8778c58cbc937c065ee148105b3580395e7650bb - fully removed temporary workaround from the code. 

Closing for now, please reopen if the same issue occurs or open a new issue for different issues
